tools/ci/check_c_api_usage.py

```diff
@@ -0,0 +1,211 @@
+#!/usr/bin/env python3
+from __future__ import annotations
+
+import argparse
+import os
+import re
+import sys
+import tempfile
+from pathlib import Path
+
+"""
+Borrow-ref C API linter (Python version).
+
+- Recursively scans source files under --root (default: numpy)
+- Matches suspicious CPython C-API calls as whole identifiers
+- Skips:
+  - lines with '// noqa: borrowed-ref OK' or
+    '// noqa: borrowed-ref - manual fix needed'
+  - line comments (// ...)
+  - block comments (/* ... */), even when they span lines
+- Prints findings and exits 1 if any issues found, else 0
+"""
+
+# List of suspicious function calls:
+SUSPICIOUS_FUNCS: tuple[str, ...] = (
+    "PyList_GetItem",
+    "PyDict_GetItem",
+    "PyDict_GetItemWithError",
+    "PyDict_GetItemString",
+    "PyDict_SetDefault",
+    "PyDict_Next",
+    "PyWeakref_GetObject",
+    "PyWeakref_GET_OBJECT",
+    "PyList_GET_ITEM",
+    "_PyDict_GetItemStringWithError",
+    "PySequence_Fast"
+)
+
+# Match any function as a standalone C identifier: (?<!\w)(NAME)(?!\w)
+FUNC_RX = re.compile(r"(?<!\w)(?:"
+                     + "|".join(map(re.escape, SUSPICIOUS_FUNCS))
+                     + r")(?!\w)")
+
+NOQA_OK = "noqa: borrowed-ref OK"
+NOQA_MANUAL = "noqa: borrowed-ref - manual fix needed"
```

Comment:
I suppose it is minor, but some of these module-level variables get used directly in functions without passing them in, which is a bit global-like. I wonder if it might make sense to pass them through. I guess it matters less for non-library code--I might be slightly tempted to just pass them through `main` and to the function that uses them from there.